### PR TITLE
bitcoin: support sending to silent payment addresses

### PR DESCRIPTION
**Pull request by benma**

### CODE REVIEW

**Comment by benma** [comment]
This  currently adds 9228 bytes to the binary, luckily less than this PR saves :smile: https://github.com/BitBoxSwiss/bitbox02-firmware/pull/1215

**Comment by benma** [comment]
Rebased. Need to double-check if payment requests to silent payment addresses work safely (an output that that is both attached to a payment request and is generated by a SP address) and add unit tests for it, or if we should disable that combination.

**Comment by benma** [comment]
Rebased.

**Comment by benma** [comment]
Host side that verifies the generated output correctness using the DLEQ proof and returns them: https://github.com/BitBoxSwiss/bitbox02-api-go/pull/105

This includes a simulator test the complete integration is tested.

**Comment by benma** [comment]
@NickeZ one way to test it (except the integration/unit tests included in this PR) is an integration test on the host side with the simulator:

    ./scripts/docker_exec.sh make -j simulator

Then in the api-go client repo at https://github.com/BitBoxSwiss/bitbox02-api-go/pull/105, run

    SIMULATOR=/path/to/bitbox02-firmware/build-build/bin/simulator go test -v -run SilentPayment ./...

```
=== RUN   TestSimulatorBTCSignSilentPayment/simulator
--- PASS: TestSimulatorBTCSignSilentPayment (0.33s)
    --- PASS: TestSimulatorBTCSignSilentPayment/simulator (0.33s)
```

File: src/rust/streaming-silent-payments/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+    if hrp != expected_hrp {
+        return Err(());
+    }
+    let witness_version = decoded_addr.remove_witness_version().unwrap();
+    if witness_version != bech32::Fe32::Q {
+        return Err(());
+    }
+
+    let data: Vec<u8> = decoded_addr.byte_iter().collect();
+    if data.len() != 66 {
```
Comment (NickeZ) [inline src/rust/streaming-silent-payments/src/lib.rs:80]: Should magic number `66` be a const?

File: src/rust/streaming-silent-payments/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        return Err(());
+    }
+    let scan_pubkey = PublicKey::from_slice(&data[..33]).map_err(|_| ())?;
```
Comment (NickeZ) [inline src/rust/streaming-silent-payments/src/lib.rs:83]: should magic number `33` be a const?

File: src/rust/streaming-silent-payments/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+    ///
+    /// Important: if the input type cannot be represented by `InputType`, the transaction must be
+    /// aborted, as other input types may influence the silent payment outputs (e.g. P2TR script
+    /// path spends, which we currently do not support).
+    pub fn add_input(
+        &mut self,
+        input_type: InputType,
+        input_key: &SecretKey,
+        prevout: bitcoin::OutPoint,
+    ) -> Result<(), ()> {
```
Comment (NickeZ) [inline src/rust/streaming-silent-payments/src/lib.rs:171]: Consider using an error type other than `()`.

File: src/keystore.h [hunk 4]
```diff
@@ -303,6 +303,12 @@ USE_RESULT bool keystore_secp256k1_schnorr_bip86_sign(
     const uint8_t* msg32,
     uint8_t* sig64_out);
 
+USE_RESULT bool keystore_secp256k1_get_private_key(
```
Comment (NickeZ) [inline src/keystore.h:306]: Missing doc comment for this function.

File: messages/btc.proto [hunk 5]
```diff
@@ -137,6 +138,9 @@ message BTCSignNextResponse {
   // Previous tx's input/output index in case of PREV_INPUT or PREV_OUTPUT, for the input at `index`.
   uint32 prev_index = 5;
   AntiKleptoSignerCommitment anti_klepto_signer_commitment = 6;
+  // Generated output. The host *must* verify its correctness using silent_payment_dleq_proof.
```
Comment (NickeZ) [inline messages/btc.proto:141]: If you write `silent_payment_dleq_proof` (with ticks `) it will be nicer formatted in the generated docs.

**Review by NickeZ — APPROVED** [review]
In general approved, left some minor comments.

File: src/rust/streaming-silent-payments/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+    pub dleq_proof: [u8; DLEQ_PROOF_SIZE],
+}
+
+pub enum Network {
+    Btc,
+    Tbtc,
+}
+
+impl Network {
+    fn sp_hrp(&self) -> &str {
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/streaming-silent-payments/src/lib.rs:40]: I'll add a comment what this is, but HRP is the name for that part of a bech32 address:

- Referred to as HRP in https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
- Used everywhere, like 
  - https://github.com/BitBoxSwiss/bitbox02-firmware/blob/master/src/rust/bitbox02-rust/src/hww/api/bitcoin/params.rs#L28
  - https://github.com/BitBoxSwiss/bitbox-wallet-app/blob/master/vendor/github.com/btcsuite/btcd/chaincfg/params.go#L414
  - etc.

File: src/rust/streaming-silent-payments/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+
+pub use bitcoin;
+use bitcoin::hashes::Hash;
+use bitcoin::secp256k1::{self, PublicKey, Scalar, Secp256k1, SecretKey, XOnlyPublicKey};
+
+use alloc::vec::Vec;
+
+const DLEQ_PROOF_SIZE: usize = 33 + 64;
+
+pub struct Output {
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/streaming-silent-payments/src/lib.rs]: Done

File: src/rust/streaming-silent-payments/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/streaming-silent-payments/src/lib.rs:45]: Added comment

File: src/rust/streaming-silent-payments/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    // We only allow one silent payment output for now. This tracks whether we've seen it.
+    output_created: bool,
+}
+
+fn calculate_t_k(ecdh_shared_secret: &PublicKey, k: u32) -> Result<SecretKey, ()> {
+    let hash = hash::SharedSecretHash::from_ecdh_and_k(ecdh_shared_secret, k).to_byte_array();
+    SecretKey::from_slice(&hash).map_err(|_| ())
+}
+
+fn decode_address(address: &str, expected_hrp: &str) -> Result<(PublicKey, PublicKey), ()> {
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/streaming-silent-payments/src/lib.rs]: Done

File: src/rust/streaming-silent-payments/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+    let mut decoded_addr =
+        bech32::primitives::decode::CheckedHrpstring::new::<bech32::Bech32m>(address)
+            .map_err(|_| ())?;
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/streaming-silent-payments/src/lib.rs:77]: Skpping for now

File: src/rust/streaming-silent-payments/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+// The host can then check that the generated output is correct by:
+// - independently computing a_sum_pubkey by summing the pubkeys of the inputs (see
+//   https://github.com/bitcoin/bips/blob/ad1d3bc2a7b0d84247c29f847e85c35283094e2f/bip-0352.mediawiki#user-content-Inputs_For_Shared_Secret_Derivation)
+//
+// - verifying using the dleq proof that the discrete log of a_sum_pubkey to G is the same as the
+//   discrete log of c_pubkey to the base scan_pubkey.
+// - re-computing the output (https://github.com/bitcoin/bips/blob/ad1d3bc2a7b0d84247c29f847e85c35283094e2f/bip-0352.mediawiki#creating-outputs)
+//   using ecdsa_shared_secret = input_hash·a_sum·scan_pubkey = input_hash·c_pubkey
+// - verifying that the created output matches the re-computed output.
+fn create_dleq_proof<C: secp256k1::Verification>(
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/streaming-silent-payments/src/lib.rs]: Done, worked.

File: src/rust/streaming-silent-payments/src/lib.rs (hunk 1 above)
Comment (benma) [reply src/rust/streaming-silent-payments/src/lib.rs]: added `const PUBKEY_LEN: usize = 33;`, and used `2 * PUBKEYLEN` here.

File: src/keystore.h (hunk 4 above)
Comment (benma) [reply src/keystore.h:314]: Added

File: messages/btc.proto (hunk 5 above)
Comment (benma) [reply messages/btc.proto]: Done

**Pull request merged**